coriolis/tests/osmorphing/osmount/test_base.py

```diff
@@ -0,0 +1,1014 @@
+            mock.call(self.base_os_mount_tools._ssh, "/tmp/tmp_dir/run"),
+        ])
+
+        expected_result = ('/tmp/tmp_dir', '/dev/sda')
+
+        self.assertEqual(result, expected_result)
+
+    @mock.patch.object(base.utils, 'test_ssh_path')
+    @mock.patch.object(base.BaseSSHOSMountTools, '_exec_cmd')
+    @mock.patch.object(base.BaseLinuxOSMountTools, '_find_dev_with_contents')
+    def test__find_and_mount_root_with_exception(
+            self, mock_find_dev_with_contents, mock_exec_cmd,
+            mock_test_ssh_path):
+        devices = ["/dev/sda", "/dev/sdb"]
+        mock_exec_cmd.return_value = b"/tmp/tmp_dir"
+        mock_find_dev_with_contents.return_value = None
+
+        self.assertRaises(exception.OperatingSystemNotFound,
+                          self.base_os_mount_tools._find_and_mount_root,
+                          devices)
+
+        mock_find_dev_with_contents.assert_called_once_with(
+            devices, all_files=self.all_files)
+        mock_exec_cmd.assert_not_called()
+        mock_test_ssh_path.assert_not_called()
+
+    @mock.patch.object(base.utils, 'test_ssh_path')
+    @mock.patch.object(base.BaseSSHOSMountTools, '_exec_cmd')
+    @mock.patch.object(base.BaseLinuxOSMountTools, '_find_dev_with_contents')
+    def test__find_and_mount_root_exec_cmd_exception(
+            self, mock_find_dev_with_contents, mock_exec_cmd,
+            mock_test_ssh_path):
+        devices = ["/dev/sda", "/dev/sdb"]
+        mock_exec_cmd.side_effect = [b"/tmp/tmp_dir", CoriolisTestException()]
+        mock_find_dev_with_contents.return_value = "/dev/sda"
+        mock_test_ssh_path.return_value = True
+
+        with self.assertLogs('coriolis.osmorphing.osmount.base',
+                             level=logging.WARN):
+            self.assertRaises(
+                CoriolisTestException,
+                self.base_os_mount_tools._find_and_mount_root,
+                devices
+            )
+        mock_exec_cmd.assert_has_calls([
+            mock.call("mktemp -d"),
+            mock.call("sudo mount /dev/sda /tmp/tmp_dir"),
+            mock.call("sudo umount /tmp/tmp_dir"),
+            mock.call("sudo rmdir /tmp/tmp_dir"),
+        ])
+
+        mock_find_dev_with_contents.assert_called_once_with(
+            devices, all_files=['etc', 'bin', 'sbin', 'boot'])
+        mock_test_ssh_path.assert_not_called()
+
+    @mock.patch.object(base.BaseSSHOSMountTools, '_exec_cmd')
+    @mock.patch.object(base.BaseLinuxOSMountTools, '_get_vgs')
+    @mock.patch.object(base.BaseLinuxOSMountTools, '_get_mounted_devices')
+    @mock.patch.object(base.BaseLinuxOSMountTools, '_find_and_mount_root')
+    @mock.patch.object(base.BaseLinuxOSMountTools, '_find_dev_with_contents')
+    @mock.patch.object(base.BaseLinuxOSMountTools, '_get_volume_block_devices')
+    @mock.patch.object(
+        base.BaseLinuxOSMountTools, '_check_mount_fstab_partitions'
+    )
+    def test_mount_os(self, mock_check_mount_fstab_partitions,
```

Comment:
I feel that this test case would be stronger if `mock_exec_cmd` had side-effects with meaningful returns for every call. I suggest this:
```python3
@mock.patch.object(base.BaseSSHOSMountTools, '_exec_cmd')
@mock.patch.object(base.BaseLinuxOSMountTools, '_get_vgs')
@mock.patch.object(base.BaseLinuxOSMountTools, '_get_mounted_devices')
@mock.patch.object(base.BaseLinuxOSMountTools, '_find_and_mount_root')
@mock.patch.object(base.BaseLinuxOSMountTools, '_find_dev_with_contents')
@mock.patch.object(base.BaseLinuxOSMountTools, '_get_volume_block_devices')
@mock.patch.object(
    base.BaseLinuxOSMountTools, '_check_mount_fstab_partitions'
)
def test_mount_os(self, mock_check_mount_fstab_partitions,
                    mock_get_volume_block_devices,
                    mock_find_dev_with_contents, mock_find_and_mount_root,
                    mock_get_mounted_devices, mock_get_vgs, mock_exec_cmd):
    mock_get_volume_block_devices.return_value = ["/dev/sda", "/dev/sdb"]
    mock_find_and_mount_root.return_value = ("/tmp/tmp_dir", "/dev/sdb1")
    mock_get_vgs.return_value = {"vg1": "/dev/sda1"}
    mock_get_mounted_devices.return_value = ["/dev/sda1"]
    mock_find_dev_with_contents.return_value = None
    mock_exec_cmd.side_effect = [
        b"",
        b"/dev/sda1",
        b"",
        b"/dev/sdb1",
        b"",
        b"",
        b"",
        b"/dev/vg1/lv1",
        b"/dev/sda1",
        b"/dev/sdb1",
        b"ext4",
        b"/dev/vg1/lv1",
        b"ext4",
        b"/dev/vg1/random-lv",
    ]

    result = self.base_os_mount_tools.mount_os()

    mock_exec_cmd.assert_has_calls([
        mock.call('sudo partx -v -a /dev/sda || true'),
        mock.call('sudo ls -1 /dev/sda*'),
        mock.call('sudo partx -v -a /dev/sdb || true'),
        mock.call('sudo ls -1 /dev/sdb*'),
        mock.call('sudo vgck'),
        mock.call('sudo vgchange -ay vg1'),
        mock.call('sudo vgchange --refresh'),
        mock.call('sudo ls -1 /dev/vg1/*'),
        mock.call('readlink -en /dev/sda1'),
        mock.call('readlink -en /dev/sdb1'),
        mock.call('sudo blkid -o value -s TYPE /dev/sdb1 || true'),
        mock.call('readlink -en /dev/vg1/lv1'),
        mock.call('sudo blkid -o value -s TYPE /dev/vg1/lv1 || true'),
        mock.call('sudo lvdisplay -c')
    ])

    mock_get_volume_block_devices.assert_called_once()
    mock_find_dev_with_contents.assert_called_once_with(
        ['/dev/sdb1', '/dev/vg1/lv1'], one_of_files=["grub", "grub2"])
    mock_get_mounted_devices.assert_called_once()
    mock_find_and_mount_root.assert_called_once()
    mock_get_vgs.assert_called_once()
    mock_check_mount_fstab_partitions.assert_called_once_with(
        "/tmp/tmp_dir", mountable_lvm_devs=['/dev/vg1/random-lv'])

    self.assertEqual(result, ('/tmp/tmp_dir', '/dev/sdb1'))

```